### ISSUE TITLE
[Rust] $CARGO_HOME/bin not in $PATH

### ISSUE THREAD

**Issue by duckinator** [gitpod-io/gitpod#587]
**tl;dr:** `cargo install` does not place executables in your `$PATH`.

### Details

The basic problem is:

1. `$CARGO_HOME` is `/workspace/.cargo`,
2. `cargo install` places executables in `$CARGO_HOME/bin`, and
2. `$PATH` does not include `$CARGO_HOME/bin`.

It is worth noting, however, that `/home/gitpod/.cargo/bin` is in PATH and should probably stay there -- it's where rustup and cargo put things [in full/Dockerfile](https://github.com/gitpod-io/workspace-images/blob/master/full/Dockerfile#L202-L207).

### Where CARGO_HOME is defined

It looks like `$CARGO_HOME` is set by `/home/gitpod/.profile`:

```
# Set CARGO_HOME to reside in workspace if:
#  - it's RUNTIME (/workspace present)
if [ -d /workspace ]; then
    export CARGO_HOME=/workspace/.cargo
fi
```

I'm not sure where this comes from, but changing that section to this should work:

```
# Set CARGO_HOME to reside in workspace if:
#  - it's RUNTIME (/workspace present)
#
# Then, if we set CARGO_HOME, update PATH.
if [ -d /workspace ]; then
    export CARGO_HOME=/workspace/.cargo
    export PATH=$CARGO_HOME/bin:$PATH
fi
```

**Comment by AlexTugarev** [comment]
@duckinator thanks a lot for reporting this! I'll take care of this.

> If the thing that appends to ~/.profile is also open-source and you point me to it I'll happily make a PR for it.

@geropl once described how to make these additions to bash profiles customizable which would simplify a lot of things. Stay tuned!

**Comment by geropl** [comment]
@duckinator Thank you for your detailed analysis! The path is indeed just missing.
> After adjusting the PATH, I discovered another problem caused by this setup:
> rustup doesn't like being installed outside of $CARGO_HOME

rustup reports this error anyway with our setup (rustup not being installed in `CARGO_HOME`), but seems to work 100% despite the error.

The current situation is a trade-off due to the way our persistence works:
 - we want dependencies (`CARGO_HOME`) be stored under `/workspace/.cargo` because that is the folder which gets persisted between workspace restarts
 - we want toolchains (`RUSTUP_HOME`) to be located under `/home/gitpod/.rustup` because:
   - they tend to get very large (~1GB per toolchain) and tend to slow down workspace startups
   - they should be (fairly) static per repo

The general recommendation for Rust is: **If you want a custom/specific toolchains (`rustup toolchain install ..`) install it in your Dockerfile** (like [here](https://github.com/geropl/ripgrep/blob/master/gitpod.Dockerfile) for example)

TODO for me:
 - add Rust section to https://docs.gitpod.io explaining this

**Comment by duckinator** [comment]
Thanks for all the info! I think I've got a working setup figured out, but I'll let y'all decide when to close this issue since it seems I did find some stuff you want to follow up on. :+1:

**Comment by AlexTugarev** [comment]
The fix is not yet deployed. 
⏲

**Comment by svenefftinge** [comment]
deployed
